### ISSUE TITLE
Naming deref target leads to incorrect lifetime errors

### ISSUE THREAD

**Issue by CAD97** [rust-lang/rust#60871]
[playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=2d6a330fff6b07b13560649dadaf1cae)

This impl compiles:

```rust
unsafe impl<D> TrustedContainer for D
where
    D::Target: TrustedContainer,
    D: ops::Deref,
{
    type Item = <D::Target as TrustedContainer>::Item;

    unsafe fn get_unchecked<'a>(&self, i: usize) -> &Self::Item {
        <D::Target>::get_unchecked(self, i)
    }
}
```

while this this theoretically identical one doesn't:

```rust
unsafe impl<T: ?Sized, D> TrustedContainer for D
where
    T: TrustedContainer,
    D: ops::Deref<Target = T>,
{
    type Item = T::Item;

    unsafe fn get_unchecked(&self, i: usize) -> &Self::Item {
        T::get_unchecked(self, i)
    }
}
```

Type context:

```rust
pub unsafe trait TrustedContainer {
    type Item: ?Sized + TrustedItem<Self>;
    
    unsafe fn get_unchecked(&self, i: usize) -> &Self::Item;
}

pub unsafe trait TrustedItem<A: TrustedContainer<Item=Self> + ?Sized> {
    type Unit;
}

unsafe impl<T: ?Sized, Array: ?Sized, D> TrustedItem<D> for T
where
    T: TrustedItem<Array>,
    Array: TrustedContainer<Item = T>,
    D: ops::Deref<Target = Array> + TrustedContainer<Item = T>,
{
    type Unit = T::Unit;
}
```

**Comment by KrishnaSannasi** [comment]
```rust
trait Trait {
    type Item: ?Sized;
    
    fn do_work(&self) -> &Self::Item;
}

trait Control {
    type Assoc: ?Sized;
    
    fn convert(&self) -> &Self::Assoc;
}

// This one doesn't
impl<T: ?Sized, D> Trait for D
where
    T: Trait,
    D: Control<Assoc = T>,
{
    type Item = T::Item;

    fn do_work(&self) -> &Self::Item {
        self.convert().do_work()
    }
}

// This one works
impl<D> Trait for D
where
    D: Control,
    D::Assoc: Trait,
{
    type Item = <D::Assoc as Trait>::Item;

    fn do_work(&self) -> &Self::Item {
        self.convert().do_work()
    }
}
```

Minimal example

**Comment by raphaelcohn** [comment]
Just had this same error today. Necrobumping to see if anyone actually cares... I'm a little cross; I've had a day of 'we haven't finished this yet' errors in the compiler.
